docs/core/misc/exceptions.md

```diff
@@ -0,0 +1,28 @@
+# Exceptions in Core
+
+From version 2.3.0 we try to follow few rules about how we use exceptions. All new
+code MUST follow these.
+
+## Usage
+
+You MAY use any exceptions in Core's logic. Exceptions from `wire.errors` SHOULD be
+the final exceptions that are thrown and SHOULD NOT be caught. Note that `wire.Error`
+is a type of exception that is intended to be sent out over the wire. It should only
+be used in contexts where that behavior is appropriate.
```

Comment:
Custom exception type hierarchies SHOULD always be derived directly from `Exception`. They SHOULD NOT be derived from other built-in exceptions (such as ValueError, TypeError, etc.)

Deriving a custom exception type signals an intention to catch and handle it somewhere in the code. For this reason, custom exception types SHOULD NOT be derived from `wire.Error` and subclasses.